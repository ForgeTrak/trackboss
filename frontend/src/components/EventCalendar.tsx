import React, { useEffect, useState, useContext } from 'react';
import moment from 'moment';
import { Calendar, DateLocalizer, Messages, momentLocalizer, View, ViewsProps } from 'react-big-calendar';
import { Text, Flex, Spacer, useDisclosure, Box } from '@chakra-ui/react';
import 'react-big-calendar/lib/css/react-big-calendar.css';
import SelectedEventModal from './SelectedEventModal';
import FamilySignUpModal from './FamilySignUpModal';
import { UserContext } from '../contexts/UserContext';
import { deleteJob, getCalendarJobs, updateJob } from '../controller/job';
import { getFamilyMembers } from '../controller/member';
import { DeletedEvent, DeleteEventResponse, Event, PostNewEventRequest } from '../../../src/typedefs/event';
import { DeletedJob, Job, PatchJobRequest } from '../../../src/typedefs/job';
import { createEvent, deleteEvent, getCalendarEvents } from '../controller/event';
import { ErrorResponse } from '../../../src/typedefs/errorResponse';
import CreateEventModal from './CreateEventModal';
import { GetMemberListResponse } from '../../../src/typedefs/member';

const RenderToolbar = require('react-big-calendar/lib/Toolbar');

const localizer: DateLocalizer = momentLocalizer(moment);

interface CustomToolbarProps {
    date: Date,
    view: View,
    views: ViewsProps<any, never>,
    label: string,
}

const customMessages: Messages = {
    next: 'Next',
    previous: 'Previous',
    today: 'Today',
    day: 'Day',
    week: 'Week',
    month: 'Month',
    agenda: 'Agenda',
};

// async function getSelectedJobAttendees(): Promise<any> {
//     const attendees = await getJobAttendees();
//     return attendees;
// }

async function getCalendarEventsLocal(token: string) {
    const events = await getCalendarEvents(token);
    const jobs = await getCalendarJobs(token);

    let calendarEvents: Array<Job | Event> = [];
    if (events && jobs) {
        calendarEvents = calendarEvents.concat(events);
        calendarEvents = calendarEvents.concat(jobs);
    }
    return calendarEvents;
}

export default function EventCalendar() {
    const { state } = useContext(UserContext);
    const { onClose: onViewEventClose, isOpen: isViewEventOpen, onOpen: onViewEventOpen } = useDisclosure();
    const { onClose: onSignUpClose, isOpen: isSignUpOpen, onOpen: onSignUpOpen } = useDisclosure();
    const [selectedEvent, setSelectedEvent] = useState<Event | Job>();
    // const [eventAttendees, setAttendees] = useState<any>();
    const [familyMembers, setFamilyMembers] = useState<any>();
    const [calendarEvents, setCalendarEvents] = useState<Array<Job | Event>>([]);

    function isEvent(event: Event | Job): event is Event {
        if ((event as Event).eventType) {
            return true;
        }
        // else, its a Job
        return false;
    }

    function deleteEventWasSuccessful(calendarEvent: DeletedEvent | ErrorResponse): calendarEvent is DeletedEvent {
        if ((calendarEvent as DeletedEvent).eventId) {
            return true;
        }
        // else, its an error
        return false;
    }

    function deleteJobWasSuccessfull(calendarEvent: DeletedJob | ErrorResponse): calendarEvent is DeletedJob {
        if ((calendarEvent as DeletedJob).jobId) {
            return true;
        }
        // else, its an error
        return false;
    }

    async function createEventLocal(newEvent: PostNewEventRequest) {
        const startDate = moment(newEvent.startDate);
        const endDate = moment(newEvent.endDate);
        // The server's expected format, ends at minutes
        newEvent.startDate = startDate.toISOString(true).slice(0, -10);
        newEvent.endDate = endDate.toISOString(true).slice(0, -10);

        const res: Event | ErrorResponse = await createEvent(state.token, newEvent);
        // eslint-disable-next-line no-console
        console.log(res); // TODO: 500 response for new event
        const test = await getCalendarEventsLocal(state.token);
        setCalendarEvents(test);
    }

    async function signUpForJob(patchInfo: { jobId: number, editedJob: PatchJobRequest }) {
        const res = await updateJob(state.token, patchInfo.jobId, patchInfo.editedJob);
        if ('reason' in res) {
            // eslint-disable-next-line no-console
            console.log(res.reason);
        } else {
            setCalendarEvents(await getCalendarEventsLocal(state.token));
        }
    }

    async function deleteEventLocal() {
        if (selectedEvent && isEvent(selectedEvent)) {
            const response: DeleteEventResponse | ErrorResponse = await deleteEvent(state.token, selectedEvent.eventId);
            if (deleteEventWasSuccessful(response)) {
                const newCalendarEvents = calendarEvents.filter((e: any) => e.eventType !== selectedEvent?.eventType);
                setCalendarEvents(newCalendarEvents);
            } else {
                // eslint-disable-next-line no-console
                console.log(response.reason);
            }
        } else if (selectedEvent) {
            const response = await deleteJob(state.token, selectedEvent.jobId);
            if (deleteJobWasSuccessfull(response)) {
                const newCalendarEvents = calendarEvents.filter((e: any) => e.jobId !== selectedEvent?.jobId);
                setCalendarEvents(newCalendarEvents);
            } else {
                // eslint-disable-next-line no-console
                console.log(response.reason);
            }
        }
    }

    async function getFamilyMembersLocal() {
        if (state.user) {
            const res: GetMemberListResponse = await getFamilyMembers(state.token, state.user.membershipId);
            if ('reason' in res) {
                console.log(res.reason);
            } else {
                return res;
            }
        }
        return undefined;
    }

    useEffect(() => {
        async function getData() {
            console.log(state.user);
            // const attendees = await getSelectedJobAttendees();
            const currentFamilyMembers = await getFamilyMembersLocal();
            // setAttendees(attendees);
            setFamilyMembers(currentFamilyMembers);
            setCalendarEvents(await getCalendarEventsLocal(state.token));
        }
        getData();
    }, []);

    return (
        <div>
            <Box pt={5} pb={5}>
                {/* eslint-disable-next-line react/jsx-no-bind */}
                <CreateEventModal createEventLocal={createEventLocal} />
            </Box>
            <Calendar
                defaultView="month"
                events={calendarEvents}
                selected={selectedEvent}
                onSelectEvent={
                    (calendarEvent) => {
                        setSelectedEvent(calendarEvent);
                        onViewEventOpen();
                    }
                }
                localizer={localizer}
                eventPropGetter={
                    (calendarEvent) => {
                        const newStyle = {
                            backgroundColor: 'lightgrey',
                            color: 'black',
                        };
                        if (isEvent(calendarEvent)) {
                            if (calendarEvent.eventType === 'Meeting') {
                                newStyle.backgroundColor = '#76CE6F'; // green
                            } else if (
                                calendarEvent.eventType === 'Yearly Job' ||
                                calendarEvent.eventType === 'Work Day') {
                                newStyle.backgroundColor = '#4B0082'; // purple
                            } else if (
                                calendarEvent.eventType === 'Race' ||
                                calendarEvent.eventType === 'Race Week' ||
                                calendarEvent.eventType === 'XO Race' ||
                                calendarEvent.eventType === 'Harescramble') {
                                newStyle.backgroundColor = '#EE6439'; // red
                            } else if (
                                calendarEvent.eventType === 'Camp and Ride' ||
                                calendarEvent.eventType === 'Ride Day') {
                                newStyle.backgroundColor = '#D3D3D3'; // lightgrey
                            } else { // its a Job
                                newStyle.backgroundColor = 'lightblue';
                            }
                            return {
                                style: newStyle,
                            };
                        }
                        // unreachable
                        return {};
                    }
                }
                style={{ height: '70vh' }}
                components={
                    {
                        toolbar: React.memo((toolbarProps: CustomToolbarProps, children) => (
                            <RenderToolbar
                                // Necessary for the custom toolbar to render properly
                                // eslint-disable-next-line react/jsx-props-no-spreading
                                {...toolbarProps}
                                localizer={
                                    {
                                        messages: {
                                            ...customMessages,
                                        },
                                    }
                                }
                                label={
                                    (
                                        <Flex pl={20}>
                                            <Spacer />
                                            <Text
                                                fontFamily="sans-serif"
                                                fontWeight={30}
                                                fontSize="3xl"
                                            >
                                                {toolbarProps.label}
                                            </Text>
                                            <Spacer />
                                        </Flex>
                                    )
                                }
                            >
                                {children}
                            </RenderToolbar>
                        )),
                    }
                }
            />
            {
                selectedEvent && (
                    <SelectedEventModal
                        isOpen={isViewEventOpen}
                        onClose={onViewEventClose}
                        selectedEvent={selectedEvent}
                        onSignUpOpen={onSignUpOpen}
                        admin={state.user?.memberType === 'Admin'}
                        // eslint-disable-next-line react/jsx-no-bind
                        deleteEvent={deleteEventLocal}
                        // eslint-disable-next-line react/jsx-no-bind
                        signUpForJob={signUpForJob}
                    />
                )
            }
            {
                familyMembers && selectedEvent && (
                    <FamilySignUpModal
                        isOpen={isSignUpOpen}
                        onClose={onSignUpClose}
                        familyMembers={familyMembers}
                        selectedEvent={selectedEvent}
                        // eslint-disable-next-line react/jsx-no-bind
                        signUpForJob={signUpForJob}
                    />
                )
            }
        </div>
    );
}
