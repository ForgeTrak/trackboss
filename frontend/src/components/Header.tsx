import React, { useContext } from 'react';
import {
    Heading,
    Flex,
    Spacer,
    Box,
    Button,
    Text,
} from '@chakra-ui/react';

import { useNavigate } from 'react-router-dom';
import HamburgerMenu from './HamburgerMenu';
import { UserContext } from '../contexts/UserContext';

interface pageProps {
    title: string;
    activeButtonId: number;
}

export default function Header(props:pageProps) {
    const navigate = useNavigate();
    const navigateToMembers = () => {
        const path = '/members';
        navigate(path);
    };

    const { state, update } = useContext(UserContext);

    return (
        <div>
            <Flex bg="white" boxShadow="lg" padding="6">
                <HamburgerMenu activeButtonId={props.activeButtonId} />
                <Spacer />
                <Box>
                    <Heading pr={90} size="3xl">{props.title}</Heading>
                </Box>
                <Spacer />
            </Flex>
            {
                state.storedUser !== undefined && (
                    <Flex height={50} backgroundColor="red">
                        <Text
                            mt={2}
                            ml={10}
                            fontStyle="bold"
                            color="white"
                            fontSize="xl"
                        >
                            Currently acting as :
                            {' '}
                            { `${state.user?.firstName} ${state.user?.lastName}`}
                        </Text>
                        <Spacer />
                        <Button
                            mt={1.5}
                            ml={15}
                            mr={10}
                            size="md"
                            variant="outline"
                            color="white"
                            onClick={
                                () => {
                                    navigateToMembers();
                                    const originalUser = state.storedUser;
                                    // Return to the original user, and clear the storedUser
                                    update({
                                        loggedIn: true,
                                        token: state.token,
                                        user: originalUser,
                                        storedUser: undefined,
                                    });
                                }
                            }
                        >
                            RETURN TO YOUR PROFILE
                        </Button>
                    </Flex>
                )
            }
        </div>
    );
}
