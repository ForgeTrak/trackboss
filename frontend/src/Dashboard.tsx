import React from 'react';
import './App.css';
import { ChakraProvider } from '@chakra-ui/react';
import theme from './theme';
import Header from './components/Header';

function Dashboard() {
    return (
        <ChakraProvider theme={theme}>
            <Header title="Dashboard" />
        </ChakraProvider>
    );
}

export default Dashboard;
