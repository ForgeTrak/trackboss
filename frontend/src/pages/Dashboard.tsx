import React, { useEffect, useState, useContext } from 'react';
import { ChakraProvider, Center, SimpleGrid, VStack } from '@chakra-ui/react';
import { UserContext } from '../contexts/UserContext';
import theme from '../theme';
import Header from '../components/Header';
import WorkPointsCard from '../components/WorkPointsCard';
import ImportantLinksCard from '../components/ImportantLinksCard';
import EventCard from '../components/EventCard';
import { getEventCardProps } from '../controller/event';
import { getWorkPointsTotal } from '../controller/workPoints';
import { getYearlyThresholdValue } from '../controller/billing';
import GreetingText from '../components/GreetingText';
import { getTodaysDate } from '../controller/utils';
import { getGateCodeLatest } from '../controller/gateCode';
import { GateCode } from '../../../src/typedefs/gateCode';

async function getEventCardPropsLocal(token: string): Promise<any | undefined> {
    const nowString = getTodaysDate();
    const props = await getEventCardProps(token, `${nowString}-`);
    return props;
}

async function getWorkPointsPercentage(token: string, memberId: number) {
    const workPoints = await getWorkPointsTotal(token, memberId);
    const threshold = await getYearlyThresholdValue(token);
    if (workPoints && threshold) {
        return Math.ceil((workPoints / threshold) * 100);
    }
    // else
    return 0;
}

function Dashboard() {
    const { state } = useContext(UserContext);
    const [eventCardProps, setEventCardProps] = useState<any>();
    const [percent, setPercent] = useState<number>(0);
    const [gateCode, setGateCode] = useState<string>('');

    useEffect(() => {
        async function getData() {
            setEventCardProps(await getEventCardPropsLocal(state.token));
            if (state.user) {
                setPercent(await getWorkPointsPercentage(state.token, state.user.memberId));
            }
        }
        getData();
    }, [state.user]);

    useEffect(() => {
        async function getData() {
            const latestGateCode = await getGateCodeLatest(state.token) as GateCode;
            setGateCode(latestGateCode.gateCode);
        }
        getData();
    });

    return (
        <ChakraProvider theme={theme}>
            <VStack align="left" spacing="2em">
                <Header title="Track Boss Dashboard" activeButtonId={1} />
                {
                    state.user && (
                        <GreetingText name={`${state.user.firstName} ${state.user.lastName}`} />
                    )
                }
                <Center>
                    <SimpleGrid columns={[1, null, 3]} spacing="20px">
                        <WorkPointsCard percent={percent} />
                        {
                            eventCardProps ? (
                                <EventCard
                                    date={eventCardProps.start}
                                    startTime={eventCardProps.time}
                                    name={eventCardProps.title}
                                />
                            ) : (
                                <EventCard
                                    date=""
                                    startTime=""
                                    name=""
                                />
                            )
                        }
                        <ImportantLinksCard gateCode={gateCode} />
                    </SimpleGrid>
                </Center>
            </VStack>
        </ChakraProvider>
    );
}

export default Dashboard;
