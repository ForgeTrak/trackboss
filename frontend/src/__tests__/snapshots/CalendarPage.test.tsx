import { Box } from '@chakra-ui/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Header from '../../components/Header';
import Dashboard from '../../pages/Dashboard';
import EventCalendar from '../../components/EventCalendar';

const upcomingEvents = [
    {
        title: 'Test Event',
        start: new Date('2022-02-08T00:10:00'),
        end: new Date('2022-02-09T00:10:00'),
        type: 'Testing',
    },
];
jest.mock('../../components/EventCalendar', () => 'EventCalendar');

describe('calendar page', () => {
    it('renders correctly', () => {
        const cal = renderer.create(
            <BrowserRouter>
                <Header title="Calendar" activeButtonId={2} />
            <Box pl={20} pr={20} pt={20}>
                <EventCalendar events={upcomingEvents} />
            </Box>
            </BrowserRouter>,
        ).toJSON();
        expect(cal).toMatchSnapshot();
    });
});
