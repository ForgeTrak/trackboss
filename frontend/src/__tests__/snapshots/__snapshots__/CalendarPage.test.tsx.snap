// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendar page renders correctly 1`] = `
Array [
  <div>
    <div
      className="css-108gr1"
    >
      <div
        className="chakra-stack css-84zodg"
      >
        <button
          aria-label="Back"
          className="chakra-button css-1izmcir"
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden={true}
            fill="currentColor"
            focusable={false}
            height="lg"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="lg"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
            />
          </svg>
        </button>
        <div>
          <button
            aria-label="Menu"
            className="chakra-button css-4exw4h"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-hidden={true}
              fill="currentColor"
              focusable={false}
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0 624H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0-312H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        className="css-17xejub"
      />
      <div
        className="css-0"
      >
        <h2
          className="chakra-heading css-10o0cdp"
        >
          Calendar
        </h2>
      </div>
      <div
        className="css-17xejub"
      />
    </div>
  </div>,
  <div
    className="css-soevlb"
  >
    <div>
      <button
        className="chakra-button css-1igmqd6"
        onClick={[Function]}
        type="button"
      >
        Create New Event
      </button>
    </div>
  </div>,
  <div
    className="css-1qqmsec"
  >
    <EventCalendar
      events={
        Array [
          Object {
            "end": 2022-02-09T05:10:00.000Z,
            "start": 2022-02-08T05:10:00.000Z,
            "title": "Test Event",
            "type": "Testing",
          },
        ]
      }
    />
  </div>,
]
`;
